File: pset2/vigenere/vigenere.c
/* !! USING THE CS50 HEADER
 *
 *
 * CS50 2017 version
 * Problem Set 2 - (http://docs.cs50.net/2017/x/psets/2/pset2.html)
 * VIGENERE: vigenere.c  - (http://docs.cs50.net/problems/vigenere/vigenere.html)
 *
 * 
 * VIGENERE:
 * This algorithm cipher text using the vigenere cipher.
 * You must ENTER the KEY (a phrase) on command line intarface (CLI)
 * otherwise a warning is printed.
 * INPUT: KEY, plaintext
 * OUTPUT: ciphertext
 *
 *
 * Usage and examples:
 * $ ./vigenere ABC
 * plaintext:  HELLO
 * ciphertext: HFNLP
 *
 * $ ./vigenere
 * You must enter the key: ./vigenere KEY
 *
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <cs50.h>

int main(int argc, string argv[])
{
    // Check command line input correctness
    if (argc != 2)
    {
        printf("You must enter the key: ./vigenere KEY\n");
        return 1;
    }
    // Every char of k must be a letter
    string k = argv[1];
    int lk = strlen(argv[1]);

    // Testing every char of k
    for (int i = 0; i < lk; i++)
    {
        // return error if a char is not a letter
        if ( !(isalpha(k[i])) )
        {
            printf("k must be alpha: ./vigenere k\n");
            return 1;
        }
    }

    // INPUT
    printf("plaintext: ");
    string p = get_string();

    // variables
    string c = p;
    int lp = strlen(p);
    int j = 0;

    // Looping through the plaintext (p)
    for (int i = 0; i < lp; i++)
    {
        // is an uppercase letter?
        if (isalpha(p[i]) && isupper(p[i]))
        {
            // When a char of k is uppercase
            if ( isupper(k[j]) )
            {
                c[i] = (( (p[i] - 'A') + (k[j] - 'A') )%26) + 'A';
            }
            else
            {
                c[i] = (( (p[i] - 'A') + (k[j] - 'a') )%26) + 'A';
            }

            // j must iterate just over index of k
            j = (j + 1) % lk;
        }

        // the char of p is letter and lowercase
        else if ( isalpha(p[i]) && islower(p[i]) )
        {
            // When a char of k is lowercase
            if (islower(k[j]))
            {
                c[i] = (( (p[i] - 'a') + (k[j] - 'a') )%26) + 'a';
            }
            else
            {
                c[i] = (( (p[i] - 'a') + (k[j] - 'A') )%26) + 'a';
            }

            // j must iterate just over index of k
            j = (j + 1) % lk;
        }
    }
    
    // OUTPUT
    printf("ciphertext: %s\n", c);
	
    // success
    return 0;
}

File: pset6/vigenere.py
"""
Pset6: Vigenere

From pset1 - vigenere
imported from C to Python

$ python vigenere.py ABC
plaintext:  HELLO
ciphertext: HFNLP

$ python vigenere.py bacon
plaintext: Meet me at the park at eleven am
ciphertext: Negh zf av huf pcfx bt gzrwep oz
"""
import sys
def main():
    # check correct number or inputs
    if len(sys.argv) != 2:
        print("You must enter the key: python vigenere.py KEY")
        exit(1)

    # every char must be alphabetical
    key = sys.argv[1]
    if (not key.isalpha()):
        print("k must be alpha: python vigenere.py k")
        exit(2)

    # input
    p = input("plaintext: ")
    c = ""
    j = 0
    lkey = len(key)

    # cipher the plaintext
    for char in p:

        # char is uppercase letter
        if char.isalpha() and char.isupper():
            # char of key is uppercase
            if key[j].isupper():
                shift = ( (ord(char)-ord('A')) + (ord(key[j])-ord('A')) )%26 + ord('A')
                c = c + chr(shift)
            else:
                shift = ( (ord(char)-ord('A')) + (ord(key[j])-ord('a')) )%26
                c = c + chr(shift + ord('A'))
            # next index for k
            j = (j+1) % lkey

        # char is lowercase letter
        elif char.isalpha() and char.islower():
            # char of key is lower
            if key[j].islower():
                shift = ( (ord(char)-ord('a')) + (ord(key[j])-ord('a')) )%26
                c = c + chr(shift + ord('a'))
            else:
                shift = ( (ord(char)-ord('a')) + (ord(key[j])-ord('A')) )%26
                c = c + chr(shift + ord('a'))
            # next index for k
            j = (j+1) % lkey

        # do not shift if char is not a letter
        else:
            c = c + char

    # print ciphered text
    print("ciphertext: {}".format(c))


if __name__ == "__main__":
    main()